### PR TITLE
Improve method `deserialize` implementation  of class ScalapbProtobufSerializer

### PR DESCRIPTION
**Pull request by carbonshow**
There are 2 ways to parse akka.util.ByteString in protobuf:
1. put all the fragments into one compact continuous memory, then protobuf parses without copy
2. just pass InputStream, protobuf does copy while parsing if it has to.
According to the test result, the first way performs better for almost all the situations. Maybe large memory fragment copy gets a better optimization.

References https://github.com/akka/akka-grpc/issues/1163

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Review by raboof — APPROVED** [review]
This seems like a good improvement for now, coupled with the data collected in the issue.

It would be *great* if we could also include the benchmarks: that way we can easily validate when upstream might have improved their implementation and we could allow more streaming again.

**Comment by raboof** [comment]
(it seems you used the 'simontao@tencent.com' email address in the commit while that hasn't been associated with your github account)

**Comment by carbonshow** [comment]
Sorry, this is my first PR on GitHub, I made some mistakes. I work at Tencent, simontao@tencent.com is my enterprise e-mail which is set globally. I have changed username and email to my GitHub account for akka-grpc. 

And I agree that including benchmark is a good idea for software evolvability and maintainability. so maybe I could do some more works：
- learn github workflow and akka-grpc  CONTRIBUTING.md
- prepare PR to include benchmark about scalapb serialization

**Comment by raboof** [comment]
> Sorry, this is my first PR on GitHub

Welcome!

> I have changed username and email to my GitHub account for akka-grpc.

The email address in the commit is still the tencent one - you can change it with something like `git commit --amend --author="Author Name <email@address.com>`

> prepare PR to include benchmark about scalapb serialization

That would be great!

I see you made the changes for this PR on the 'master' branch in your repository. If you want to be able to track multiple contributions in parallel, it is often easier to let your 'master' just follow our upstream 'master', and make changes on new branches. Perhaps the easiest way right now would be to first wait for this PR to be merged (after you update the author field on the commit), and then reset your master and start using branches.

**Comment by raboof** [comment]
OK TO TEST

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by carbonshow** [comment]
hi, raboof. I try to change the commit author following your instruction:
1. git commit --amend --author="carbonshow <wenzhitao515@qq.com>"
2. git pull
3. git push
but lightbend-cla-validator still says "at least on pull request committer is not linked to a user." Did I miss something?

And, I create a branch for scalapb deserialization benchmark instead of using master directly. After reading the code I have some questions:
- Is there specification for adding new benchmark?
- should new benchmark class extends from ```CommonBenchmark``` ?
- ```HandlerProcessingBenchmark``` is designed for measuring the whole rpc process performance. Is it necessary to add the benchmark only for deserialization？
- If the benchmark for specified part is needed, where to locate the schema? maybe under benchmarks project?

**Comment by lightbend-cla-validator** [comment]
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

**Comment by carbonshow** [comment]
> The email address in the commit is still the tencent one - you can change it with something like `git commit --amend --author="Author Name <email@address.com>`

Finally, I fix it.

> I see you made the changes for this PR on the 'master' branch in your repository. If you want to be able to track multiple contributions in parallel, it is often easier to let your 'master' just follow our upstream 'master', and make changes on new branches. Perhaps the easiest way right now would be to first wait for this PR to be merged (after you update the author field on the commit), and then reset your master and start using branches.

OK, after navigating the benchmarks project, I want to know if there is specification for adding new benchmark?
- should new benchmark class extends from ```CommonBenchmark``` ?
- ```HandlerProcessingBenchmark``` is designed for measuring the whole rpc process performance. Is it necessary to add the benchmark only for deserialization？
- If the benchmark for specified part is needed, where to locate the schema? maybe under benchmarks project?

**Review by raboof — APPROVED** [review]

**Comment by raboof** [comment]
> > The email address in the commit is still the tencent one - you can change it with something like `git commit --amend --author="Author Name <email@address.com>`
> 
> Finally, I fix it.

Great! Thanks, merged.
 
> > I see you made the changes for this PR on the 'master' branch in your repository. If you want to be able to track multiple contributions in parallel, it is often easier to let your 'master' just follow our upstream 'master', and make changes on new branches. Perhaps the easiest way right now would be to first wait for this PR to be merged (after you update the author field on the commit), and then reset your master and start using branches.
> 
> OK, after navigating the benchmarks project, I want to know if there is specification for adding new benchmark?
> 
>     * should new benchmark class extends from `CommonBenchmark` ?

I don't think it is required (but it might be helpful)

>     * `HandlerProcessingBenchmark` is designed for measuring the whole rpc process performance. Is it necessary to add the benchmark only for deserialization?

I think it would be useful to be able to benchmark that in isolation, yes

>     * If the benchmark for specified part is needed, where to locate the schema? maybe under benchmarks project?

Hmm, the benchmarks project already has generates code for protobuf, but it generates Java code. Having both Java and Scala code there might be tricky - you might want to add the `flat_package` code generator setting to avoid name clashes.

**Pull request merged**
